src/System/Random/Internal.hs

```diff
@@ -709,34 +713,47 @@ instance UniformRange Bool where
 -- | See /Floating point number caveats/ in "System.Random.Stateful".
 instance UniformRange Double where
   uniformRM (l, h) g = do
-    w64 <- uniformWord64 g
-    let x = word64ToDoubleInUnitInterval w64
+    x <- uniformDouble01M g
     return $ (h - l) * x + l
 
--- | Turns a given uniformly distributed 'Word64' value into a uniformly
--- distributed 'Double' value in the range [0, 1].
-word64ToDoubleInUnitInterval :: Word64 -> Double
-word64ToDoubleInUnitInterval w64 = d / m
+-- | Generate uniformly distributed 'Double' in the range [0, 1].
+uniformDouble01M :: StatefulGen g m => g -> m Double
+uniformDouble01M g = do
+  w64 <- uniformWord64 g
+  return $ fromIntegral w64 / m
   where
-    d = fromIntegral w64 :: Double
     m = fromIntegral (maxBound :: Word64) :: Double
-{-# INLINE word64ToDoubleInUnitInterval #-}
+
+-- | Generate uniformly distributed 'Double' in the range (0, 1]. That
+--   is result is guaranteed to be positive
+uniformDoublePos01M :: StatefulGen g m => g -> m Double
+uniformDoublePos01M g = (+ d) <$> uniformDouble01M g
+  where
+    -- We add small constant to shift generated value from zero. It's
+    -- selected as 1/2 of smallest possible nonzero value
+    d = 2.710505431213761e-20 -- 2**(-65)
```

Comment:
This is the largest value that can be added to 1.0 such that the result is still 1.0. There are many smaller values that also have this property and that would also result in never getting 0 as a result of this function. My guess is that the goal here was to avoid "excessively small values" rather than just avoiding zero, is that correct? If so, I would suggest extending the description a little to justify why this is `2^-64` and not e.g. `5e-324`, which would also avoid getting zero but would not exclude very small numbers from being returned.

Reply:
It' good question. 

No. It's 1/2 of smallest non zero value that this algorithm could generate. Largest ε that 1+ε=1 is 2^{-53}. Choice is somewhat arbitrary. Another possibility is to pick ε*2^{-64}. That is largest number that will leave all other values unchanged.

Reply:
Right, thanks for the correction.

I guess the question is: under what circumstances do people want to avoid getting a zero here, and under those circumstances, would they also want to avoid getting extremely small values?

My default here would be to exclude only the zero. Everything else should have a good reason IMO, since excluding more numbers results in a larger effect on the outcome distribution.

Reply:
> 1/2 of smallest possible nonzero value

I don't think that's right. The smallest positive double, or the smallest absolute value that is non-zero, is `5e-324`. So "1/2 of smallest possible nonzero value" would be `2.5e-324`.

I don't have a very strong opinion on what this number should be, but the comment should be updated.

Reply:
I think I can explain that what @Shimuuar suggests is actually very good. Suppose you have floating point numbers with only 2 bits for the mantissa and 2 bits for the exponent then using the proposed algorithm:

```
word32ToFloatInUnitInterval :: Int -> Float
word32ToFloatInUnitInterval w32 = f / m
  where
    f = fromIntegral w32 :: Float
    m = fromIntegral (maxBound :: Int) :: Float
```

and manually calculating for our tiny floating point values

```
map ((/4) . fromIntegral) [0,1,2,3]
[0.0,0.25,0.5,0.75]
```

The smallest value the RNG algorithm can generate is 0.25 and half of this is 0.125 so we will get

```
map ((+(0.25/2)) . (/4) . fromIntegral) [0,1,2,3]
[0.125,0.375,0.625,0.875]
```

which has the nice property that it is symmetrical in the interval. N.B. that these really are available in the floating point representation

```
normals :: [Double]
normals = [ 1/2^e * (1 + m0 * (1/2^1) + m1 * (1/2^2)) | e <- [2,1], m0 <- [0,1], m1 <- [0,1] ]
[0.25,0.3125,0.375,0.4375,0.5,0.625,0.75,0.875]
```

so no rounding is happening (that was what I was worried about).

Reply:
Of course this nice property doesn't really manifest itself with `Double` and even `Float` - you would have to sample a *lot* of `Double` to notice the mean is now correct :)

Reply:
Ah okay that clears it up.

> 1/2 of smallest possible nonzero value

"smallest possible" here does not refer to the IEEE 754 encoding but to the division algorithm we use to generate doubles. Cool. Let's put that in the comment.

For concreteness' sake, here is my suggestion:

Add a small constant shift away from zero. This constant is 1/2 of the smallest positive value generated by 'uniformDouble01M'.

(And something similar for Float).

Reply:
Ping